source/server/admin/stats_render.cc

```diff
@@ -0,0 +1,316 @@
+#include "source/server/admin/stats_render.h"
+
+#include "source/common/html/utility.h"
+#include "source/common/stats/histogram_impl.h"
+
+namespace {
+// This value found by iterating in stats_handler_speed_test.cc, maximizing the
+// performance of the Json tests, and then rounding to a power of 2.
+constexpr uint64_t JsonStatsFlushCount = 64;
+} // namespace
+namespace Envoy {
+
+using ProtoMap = Protobuf::Map<std::string, ProtobufWkt::Value>;
+
+namespace Server {
+
+void StatsTextRender::generate(Buffer::Instance& response, const std::string& name,
+                               uint64_t value) {
+  response.addFragments({name, ": ", absl::StrCat(value), "\n"});
+}
+
+void StatsTextRender::generate(Buffer::Instance& response, const std::string& name,
+                               const std::string& value) {
+  response.addFragments({name, ": \"", Html::Utility::sanitize(value), "\"\n"});
+}
+
+void StatsTextRender::generate(Buffer::Instance& response, const std::string& name,
+                               const Stats::ParentHistogram& histogram) {
+  switch (histogram_buckets_mode_) {
+  case Utility::HistogramBucketsMode::NoBuckets:
+    response.addFragments({name, ": ", histogram.quantileSummary(), "\n"});
+    break;
+  case Utility::HistogramBucketsMode::Cumulative:
+    response.addFragments({name, ": ", histogram.bucketSummary(), "\n"});
+    break;
+  case Utility::HistogramBucketsMode::Disjoint:
+    addDisjointBuckets(name, histogram, response);
+    break;
+  }
+}
+
+void StatsTextRender::finalize(Buffer::Instance&) {}
+
+// Computes disjoint buckets as text and adds them to the response buffer.
+void StatsTextRender::addDisjointBuckets(const std::string& name,
+                                         const Stats::ParentHistogram& histogram,
+                                         Buffer::Instance& response) {
+  if (!histogram.used()) {
+    response.addFragments({name, ": No recorded values\n"});
+    return;
+  }
+  response.addFragments({name, ": "});
+  std::vector<absl::string_view> bucket_summary;
+
+  const Stats::HistogramStatistics& interval_statistics = histogram.intervalStatistics();
+  Stats::ConstSupportedBuckets& supported_buckets = interval_statistics.supportedBuckets();
+  const UInt64Vec disjoint_interval_buckets = interval_statistics.computeDisjointBuckets();
+  const UInt64Vec disjoint_cumulative_buckets =
+      histogram.cumulativeStatistics().computeDisjointBuckets();
+  // Make sure all vectors are the same size.
+  ASSERT(disjoint_interval_buckets.size() == disjoint_cumulative_buckets.size());
+  ASSERT(disjoint_cumulative_buckets.size() == supported_buckets.size());
```

Comment:
I added the nit because of the following line: https://github.com/envoyproxy/envoy/pull/19693/files/9fc63d09e804bb2a9d4d3598aabc80072207e0f1#diff-fe266c26e0c63776327c17fb75338ddcc7eab67e380c0618cd6bd0260bddf80dR63

Personally, I feel it's better to have either have a RELEASE_ASSERT without the call to std::min OR ENVOY_BUG and the call to std::min, though I don't know what the standard practice is.

My reasoning is that it's probably not catastrophic if the invariant fails, so maybe we don't want to kill the Envoy if it does, but we should at least know of the failure.

Reply:
OK -- I'm not convinced in this case -- but anyway I don't want to change that in this PR. This code really was just moved from its previous location.

IMO this is better as the ASSERTs are really more for the comfort of the reader, as they are verifying properties that are guaranteed at the source of this info.